File: python/TaskInterfaceServerTest.py
#!/usr/bin/env python

import unittest
import datetime
import tempfile
import os

from TaskInterfaceServer import TaskInterfaceServer
from Task import Task
from TaskError import TaskError


class TaskInterfaceServerTest(unittest.TestCase):

	format = "%Y-%m-%d %H:%M:%S"

	def setUp(self):
		# connect to Redis database
		self.tdf = TaskInterfaceServer('localhost', 6379, 15)

		# clear the test database
		self.tdf.redis.flushall()


	def test_get_missing_task(self):
		self.assertFalse(self.tdf.getTask('asd', -1))


	def test_namesapces(self):
		# the initial list of namespaces should be empty
		self.assertFalse(self.tdf.getNamespaces())

		# add two namespaces there should be two namespaces in the list
		self.tdf.addNamespace('foo')
		self.tdf.addNamespace('bar')
		self.assertTrue(len(self.tdf.getNamespaces()) == 2)

		# change the index counter
		self.assertTrue(self.tdf.redis.get('tdf.foo.index'), '-1')
		self.tdf.redis.set('tdf.foo.index', '1')

		# try to add the namespace "foo" again it should not be added
		self.tdf.addNamespace('foo')
		self.assertTrue(len(self.tdf.getNamespaces()) == 2)

		# index counter should still be "1"
		self.assertTrue(self.tdf.redis.get('tdf.foo.index'), '1')

		# add task to "bar"
		index = self.tdf.addTask('bar', Task({'worker': 'foo'}))
		self.assertTrue(self.tdf.countQueuedTasks('bar') == 1)

		# remove the namespace "bar", now there should be only one namespace left
		self.tdf.deleteNamespace('bar')
		self.assertTrue(len(self.tdf.getNamespaces()) == 1)

		# the task added to "bar" should not exist anymore
		self.assertTrue(self.tdf.countQueuedTasks('bar') == 0)


	def test_add_and_remove_task(self):
		# create task object
		task = Task()
		runBefore = datetime.datetime.today() + datetime.timedelta(hours=1)
		# time attributes need to be a string with format yyyy-MM-dd hh:mm:ss
		task.set('runBefore', runBefore.strftime(self.format))

		# try to add without worker
		with self.assertRaises(TaskError):
			self.tdf.addTask('foo', task)

		# add task to queue
		task.set('worker', 'worker-uri')
		index = self.tdf.addTask('foo', task)
		self.assertEquals(self.tdf.getTask('foo', index).get('worker'), 'worker-uri')

		# move task to running
		index = self.tdf.redis.lpop('tdf.foo.queuing')
		self.tdf.redis.sadd('tdf.foo.running', index)
		self.assertTrue(self.tdf.countRunningTasks('foo') == 1)

		# move task to completed
		self.tdf.redis.smove('tdf.foo.running', 'tdf.foo.completed', index)
		self.assertTrue(self.tdf.countCompletedTasks('foo') == 1)

		# process task
		self.tdf.processTask('foo', index)
		self.assertTrue(self.tdf.countCompletedTasks('foo') == 0)
		self.assertTrue(self.tdf.countProcessedTasks('foo') == 1)

		# remove task
		self.assertTrue(self.tdf.deleteAllTasks('foo') == 1)
		self.assertTrue(self.tdf.countProcessedTasks('foo') == 0)
		self.assertFalse(self.tdf.getTask('foo', index))


	def test_delete_queued_tasks(self):
		# create task
		task = Task({'worker': 'worker-uri'})

		# add it to queued list
		index = self.tdf.addTask('foo', task)
		self.assertEqual(self.tdf.countQueuedTasks('foo'), 1)

		# delete all queued tasks, list should be empty and task deleted
		self.tdf.deleteQueuedTasks('foo')
		self.assertEqual(self.tdf.countQueuedTasks('foo'), 0)


	def test_delete_running_tasks(self):
		# create task
		task = Task({'worker': 'worker-uri'})
		index = self.tdf.addTask('foo', task)

		# move task to running
		self.tdf.redis.lpop('tdf.foo.queuing')
		self.tdf.redis.sadd('tdf.foo.running', index)
		self.assertEqual(self.tdf.countRunningTasks('foo'), 1)

		# delete all running tasks, list should be empty and task deleted
		self.tdf.deleteRunningTasks('foo')
		self.assertEqual(self.tdf.countRunningTasks('foo'), 0)


	def test_delete_completed_tasks(self):
		# create task
		task = Task({'worker': 'worker-uri'})
		index = self.tdf.addTask('foo', task)

		# move task to completed
		self.tdf.redis.lpop('tdf.foo.queuing')
		self.tdf.redis.sadd('tdf.foo.completed', index)
		self.assertEqual(self.tdf.countCompletedTasks('foo'), 1)

		# delete all completed tasks, list should be empty and task deleted
		self.tdf.deleteCompletedTasks('foo')
		self.assertEqual(self.tdf.countCompletedTasks('foo'), 0)


	def test_delete_processed_tasks(self):
		# create task
		task = Task({'worker': 'worker-uri'})
		index = self.tdf.addTask('foo', task)

		# move task to processed
		self.tdf.redis.lpop('tdf.foo.queuing')
		self.tdf.redis.sadd('tdf.foo.processed', index)
		self.assertEqual(self.tdf.countProcessedTasks('foo'), 1)

		# delete all processed tasks, list should be empty and task deleted
		self.tdf.deleteProcessedTasks('foo')
		self.assertEqual(self.tdf.countProcessedTasks('foo'), 0)


	def test_rescheduling(self):
		# create task and "start" it
		task = Task({'worker': 'worker-uri', 'timeout': '100000'})

		# add task to queue
		index = self.tdf.addTask('foo', task)

		# move task to running
		self.tdf.redis.lpop('tdf.foo.queuing')
		self.tdf.redis.sadd('tdf.foo.running', index)
		self.tdf.redis.hset("tdf.foo.task.%s" % index, 'started', datetime.datetime.now().strftime(self.format))

		# requeue, task should still be in the running set
		self.tdf.requeue('foo')
		self.assertEqual(self.tdf.countRunningTasks('foo'), 1)

		# change requeue time
		self.tdf.redis.hset("tdf.foo.task.%s" % index, 'timeout', -10000)

		# requeue, task should be in the queuing list
		self.tdf.requeue('foo')
		self.assertEqual(self.tdf.countRunningTasks('foo'), 0)
		self.assertEqual(self.tdf.countQueuedTasks('foo'), 1)

		# change expire timestamp
		runBefore = datetime.datetime.now() - datetime.timedelta(minutes=10)
		self.tdf.redis.hset("tdf.foo.task.%s" % index, 'runBefore', runBefore.strftime(self.format))

		# requeue (that includes checking for expired tasks)
		# task should still be in the queuing list
		self.tdf.deleteAllExpiredTasks('foo')
		self.assertEqual(self.tdf.countQueuedTasks('foo'), 0)


	def test_expire_running_tasks(self):
		# create task and "start" it
		runBefore = datetime.datetime.now() + datetime.timedelta(minutes=10)
		task = Task({'worker': 'worker-uri', 'runBefore': runBefore.strftime(self.format)})

		# add task to queue
		index = self.tdf.addTask('foo', task)

		# move task to running
		self.tdf.redis.lpop('tdf.foo.queuing')
		self.tdf.redis.sadd('tdf.foo.running', index)
		self.tdf.redis.hset("tdf.foo.task.%s" % index, 'started', datetime.datetime.now().strftime(self.format));

		# requeue (that includes checking for expired tasks)
		# task should still be in the running set
		self.tdf.requeue('foo')
		self.assertEqual(self.tdf.countRunningTasks('foo'), 1)

		# change expire timestamp
		runBefore = datetime.datetime.now() - datetime.timedelta(minutes=10)
		self.tdf.redis.hset("tdf.foo.task.%s" % index, 'runBefore', runBefore.strftime(self.format))

		# requeue, running set should be empty
		self.tdf.deleteAllExpiredTasks('foo')
		self.assertEqual(self.tdf.countRunningTasks('foo'), 0)


	def test_export_processed_tasks(self):
		tempDir = tempfile.mkdtemp()

		# create task
		task = Task({'worker': 'http://dl.dropbox.com/u/1721583/test-worker-1.0.0.zip', 'input': 'input'})
		index = self.tdf.addTask('foo', task)

		# set task processed
		self.tdf.redis.sadd('tdf.foo.processed', index)
		self.assertTrue(self.tdf.countProcessedTasks('foo') == 1)

		# export
		self.tdf.exportProcessedTasks('foo', tempDir)
		self.assertTrue(self.tdf.countProcessedTasks('foo') == 0)

		taskIndexDir = os.path.join(tempDir, 'foo', 'test-worker-1.0.0', 'tasks', datetime.datetime.today().strftime("%Y-%m-%d"), index.__str__())
		self.assertEquals(len(os.listdir(taskIndexDir)), 5)


if (__name__ == '__main__'):
	unittest.main()

File: java/server/src/test/java/de/tuda/p2p/tdf/server/test/TaskInterfaceServerTest.java
package de.tuda.p2p.tdf.server.test;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.not;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.Properties;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;
import org.junit.BeforeClass;
import org.junit.Test;

import de.tuda.p2p.tdf.common.ClientTask;
import de.tuda.p2p.tdf.common.Task;
import de.tuda.p2p.tdf.server.TaskInterfaceServer;

public class TaskInterfaceServerTest {

	private static TaskInterfaceServer taskInterface;

	DateTimeFormatter formatter = DateTimeFormat
			.forPattern("yyyy-MM-dd HH:mm:ss");

	/**
	 * Returns the database configuration
	 *
	 * @return The property object with the database configuration information
	 * @throws URISyntaxException
	 *             Thrown if the path to db.properties is not correct
	 * @throws FileNotFoundException
	 *             Thrown if the configuration file cannot be found
	 * @throws IOException
	 *             Thrown if there was an error while loading the configuration
	 */
	private static Properties getDatabaseConfig() throws URISyntaxException,
			FileNotFoundException, IOException {
		URL configUrl = TaskInterfaceServerTest.class
				.getResource("/db.properties");
		File configFile = new File(configUrl.toURI());

		Properties config = new Properties();
		config.load(new FileInputStream(configFile));

		return config;
	}

	@BeforeClass
	public static void connectToDatabase() throws URISyntaxException,
			FileNotFoundException, IOException {
		// get the database configuration from db.properties
		Properties config = getDatabaseConfig();

		String host = config.getProperty("redis.host");
		Integer port = Integer.valueOf(config.getProperty("redis.port"));
		Integer index = Integer.valueOf(config.getProperty("redis.index"));

		// connect to the test database
		taskInterface = new TaskInterfaceServer(host, port, index);

		// flush the test database
		taskInterface.getJedis().flushAll();
	}

	@Test
	public void testNamespaces() {
		// the initial list of namespaces should be empty
		assertThat(taskInterface.getNamespaces().isEmpty(), is(true));

		// add two namespaces; there should be two namespaces in the list
		taskInterface.addNamespace("foo");
		taskInterface.addNamespace("bar");
		assertThat(taskInterface.getNamespaces().size(), is(2));

		// change the index counter
		assertThat(taskInterface.getJedis().get("tdf.foo.index"), is("-1"));
		taskInterface.getJedis().set("tdf.foo.index", "1");

		// try to add the namespace "foo" again; it should not be added
		taskInterface.addNamespace("foo");
		assertThat(taskInterface.getNamespaces().size(), is(2));

		// index counter should still be "1"
		assertThat(taskInterface.getJedis().get("tdf.foo.index"), is("1"));

		// add task to "bar"
		Long index = taskInterface.addTask("bar", new Task("asd", "sdf", null, null, null, null, null, null, null));
		assertThat(taskInterface.countQueuingTasks("bar").intValue(), is(1));

		// remove the namespace "bar"; now there should be only one namespace
		// left
		taskInterface.deleteNamespace("bar");
		assertThat(taskInterface.getNamespaces().size(), is(1));

		// the task added to "bar" should not exist anymore
		assertThat(taskInterface.getTask("bar", index), equalTo(null));
	}

	@Test
	public void testAddAndRemoveTask() {
		// create task object
		Task task = new Task();
		task.setRunBefore(DateTime.now().plusHours(1));

		// try to add without worker
		Long index = taskInterface.addTask("foo", task);
		assertThat(index, is(-1L));
		task.setWorker("worker-uri");

		// add task to queue
		index = taskInterface.addTask("foo", task);
		assertThat(taskInterface.getQueuingTasks("foo").get(0).getWorker(),
				is("worker-uri"));

		// move task to running
		String idx = taskInterface.getJedis().lpop("tdf.foo.queuing");
		assertThat(idx, equalTo(index.toString()));
		taskInterface.getJedis().sadd("tdf.foo.running", idx);
		assertThat(taskInterface.countRunningTasks("foo").intValue(), is(1));

		// move task to completed
		taskInterface.getJedis().smove("tdf.foo.running", "tdf.foo.completed",
				idx);
		assertThat(taskInterface.countCompletedTasks("foo").intValue(), is(1));

		// process task
		taskInterface.processTask("foo", index);
		assertThat(taskInterface.countCompletedTasks("foo").intValue(), is(0));
		assertThat(taskInterface.countProcessedTasks("foo").intValue(), is(1));

		// update task
		task.setWorker("new-worker-uri");
		task.save(taskInterface.getJedis(), "foo", index);
		assertThat(taskInterface.getTask("foo", index).getWorker(),
				is("new-worker-uri"));

		// remove task
		assertThat(taskInterface.deleteTask("foo", index), is(true));
		assertThat(taskInterface.countProcessedTasks("foo").intValue(), is(0));

		assertThat(taskInterface.getTask("foo", index), equalTo(null));
	}

	@Test
	public void testDeleteQueuingTasks() {
		// create task
		Task task = new Task();
		task.setWorker("worker-uri");

		// add it to queuing list
		Long index = taskInterface.addTask("foo", task);
		assertThat(taskInterface.countQueuingTasks("foo").intValue(), is(1));

		// delete all queuing tasks, list should be empty and task deleted
		assertThat(taskInterface.deleteQueuingTasks("foo").intValue(), is(1));
		assertThat(taskInterface.countQueuingTasks("foo").intValue(), is(0));
		assertThat(taskInterface.getTask("foo", index), equalTo(null));
	}

	@Test
	public void testDeleteRunningTasks() {
		// create task
		Task task = new Task();
		task.setWorker("worker-uri");

		// move it to the running set
		Long index = taskInterface.addTask("foo", task);
		String idx = taskInterface.getJedis().lpop("tdf.foo.queuing");
		taskInterface.getJedis().sadd("tdf.foo.running", idx);
		assertThat(taskInterface.countRunningTasks("foo").intValue(), is(1));

		// delete all running tasks, list should be empty and task deleted
		assertThat(taskInterface.deleteRunningTasks("foo").intValue(), is(1));
		assertThat(taskInterface.countRunningTasks("foo").intValue(), is(0));
		assertThat(taskInterface.getTask("foo", index), equalTo(null));
	}

	@Test
	public void testDeleteCompletedTasks() {
		// create task
		Task task = new Task();
		task.setWorker("worker-uri");

		// move it to the completed set
		Long index = taskInterface.addTask("foo", task);
		String idx = taskInterface.getJedis().lpop("tdf.foo.queuing");
		taskInterface.getJedis().sadd("tdf.foo.completed", idx);
		assertThat(taskInterface.countCompletedTasks("foo").intValue(), is(1));

		// delete all completed tasks, list should be empty and task deleted
		assertThat(taskInterface.deleteCompletedTasks("foo").intValue(), is(1));
		assertThat(taskInterface.countCompletedTasks("foo").intValue(), is(0));
		assertThat(taskInterface.getTask("foo", index), equalTo(null));
	}

	@Test
	public void testDeleteProcessedTasks() {
		// create task
		Task task = new Task();
		task.setWorker("worker-uri");

		// move it to the processed set
		Long index = taskInterface.addTask("foo", task);
		String idx = taskInterface.getJedis().lpop("tdf.foo.queuing");
		taskInterface.getJedis().sadd("tdf.foo.processed", idx);
		assertThat(taskInterface.countProcessedTasks("foo").intValue(), is(1));

		// delete all processed tasks, list should be empty and task deleted
		assertThat(taskInterface.deleteProcessedTasks("foo").intValue(), is(1));
		assertThat(taskInterface.countProcessedTasks("foo").intValue(), is(0));
		assertThat(taskInterface.getTask("foo", index), equalTo(null));
	}

	@Test
	public void testAllTasks() {
		// create task
		Task task = new Task();
		task.setWorker("worker-uri");

		// move it to the processed set
		Long index = taskInterface.addTask("foo", task);
		String idx = taskInterface.getJedis().lpop("tdf.foo.queuing");
		taskInterface.getJedis().sadd("tdf.foo.processed", idx);
		assertThat(taskInterface.countProcessedTasks("foo").intValue(), is(1));

		// delete all tasks, processed list should be empty and task deleted
		assertThat(taskInterface.deleteAllTasks("foo").intValue(), is(1));
		assertThat(taskInterface.countProcessedTasks("foo").intValue(), is(0));
		assertThat(taskInterface.getTask("foo", index), equalTo(null));
	}

	@Test
	public void testDeleteMissingTasks() {
		// add index to queuing
		taskInterface.getJedis().lpush("tdf.foo.queuing", "42");
		assertThat(taskInterface.countQueuingTasks("foo").intValue(), is(1));

		// delete missing tasks
		assertThat(taskInterface.deleteMissingTasks("foo").intValue(), is(1));
		assertThat(taskInterface.countQueuingTasks("foo").intValue(), is(0));
	}

	@Test
	public void testRescheduling() {
		// create task and "start" it
		Task task = new Task();
		task.setWorker("worker-uri");
		task.setTimeout("100000");

		// add task to queue
		Long index = taskInterface.addTask("foo", task);

		// move task to running
		String idx = taskInterface.getJedis().lpop("tdf.foo.queuing");
		assertThat(idx, equalTo(index.toString()));
		taskInterface.getJedis().sadd("tdf.foo.running", idx);
		taskInterface.getJedis().hset("tdf.foo.task." + idx, "started",
				DateTime.now().toString(formatter));

		// requeue, task should still be in the running set
		assertThat(taskInterface.requeue("foo").intValue(), is(0));
		assertThat(taskInterface.countRunningTasks("foo").intValue(), is(1));

		// change requeue time
		taskInterface.getJedis().hset("tdf.foo.task." + idx, "timeout", "-1");

		// requeue, task should be in the queuing list
		assertThat(taskInterface.requeue("foo").intValue(), is(1));
		assertThat(taskInterface.countRunningTasks("foo").intValue(), is(0));
		assertThat(taskInterface.countQueuingTasks("foo").intValue(), is(1));

		// change expire timestamp
		taskInterface.getJedis().hset("tdf.foo.task." + index, "runBefore",
				DateTime.now().minusMinutes(10).toString(formatter));

		// requeue (that includes checking for expired tasks)
		// task should still be in the queuing list
		assertThat(taskInterface.deleteAllExpiredTasks("foo").intValue(), is(1));
		assertThat(taskInterface.countQueuingTasks("foo").intValue(), is(0));
	}

	@Test
	public void testExpireRunningTasks() {
		// create task and "start" it
		Task task = new Task();
		task.setWorker("worker-uri");
		task.setRunBefore(DateTime.now().plusMinutes(10));

		// add task to queue
		Long index = taskInterface.addTask("foo", task);

		// move task to running
		String idx = taskInterface.getJedis().lpop("tdf.foo.queuing");
		assertThat(idx, equalTo(index.toString()));
		taskInterface.getJedis().sadd("tdf.foo.running", idx);
		taskInterface.getJedis().hset("tdf.foo.task." + idx, "started",
				DateTime.now().toString(formatter));

		// requeue (that includes checking for expired tasks)
		// task should still be in the running set
		assertThat(taskInterface.requeue("foo").intValue(), is(0));
		assertThat(taskInterface.countRunningTasks("foo").intValue(), is(1));

		// change expire timestamp
		taskInterface.getJedis().hset("tdf.foo.task." + idx, "runBefore",
				DateTime.now().minusMinutes(10).toString(formatter));

		// requeue, running set should be empty
		assertThat(taskInterface.deleteAllExpiredTasks("foo").intValue(), is(1));
		assertThat(taskInterface.countRunningTasks("foo").intValue(), is(0));
	}

	@Test
	public void testExportTask() throws IOException {
		File directory = createTempDirectory();

		ClientTask task = new ClientTask();
		task.setWorker("worker-uri");
		task.setInput("input");
		task.setNamespace("foo");
		task.setBaseDir(directory);
		Long index = taskInterface.addTask("foo", task);
		assertThat(index.intValue(), is(not(-1)));

		taskInterface.getJedis().sadd("tdf.foo.processed", index.toString());
		assertThat(taskInterface.countProcessedTasks("foo").intValue(), is(1));

		assertThat(taskInterface.exportProcessedTasks("foo", directory, true, true, true, true, true).intValue(), is(1));
		assertThat(taskInterface.countProcessedTasks("foo").intValue(), is(0));

		assertThat(task.getTaskIndexDir().listFiles().length, is(5));
	}

	private File createTempDirectory() throws IOException {
		File temp = File
				.createTempFile("tmp", Long.toString(System.nanoTime()));
		;
		if (!temp.delete()) {
			throw new IOException("Could not delete temp file: "
					+ temp.getCanonicalPath());
		}

		if (!temp.mkdirs()) {
			throw new IOException("Could not create temp directory: "
					+ temp.getCanonicalPath());
		}

		System.out.println("Using temp dir: " + temp.getCanonicalPath());
		return temp;
	}

}
